### PR TITLE
[Proposal] Align API with Trainer

### PR DESCRIPTION
**Pull request by julien-c**
I've created a new `trainer_tf_barebones.py` rather than modifying `trainer_tf.py`, but it might be more helpful to look at the diff between the two.

If you need a one-line-of-code way to launch a training, like you have in your proposal, I would suggest building a wrapper around the kind of boilerplate that's in the `tf_glue_barebones.py` script – not sure of the name, maybe `TrainingLauncher` or `TrainingManager` or something

### CODE REVIEW

File: tf_training_args.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+# I am not sure, but maybe this shouldn't even be a child subclass,
+# but rather update TrainingArguments with some additional TF helpers.
+# Not sure about this...
+
+@dataclass
+class TFTrainingArguments(TrainingArguments):
+    strategy_name: str = field(default="onedevice")
+    optimizer_name: str = field(default="adam")
+    mode: str = field(default="classification")
```
Comment (julien-c) [inline tf_training_args.py:14]: Could probably be an enum

**Comment by jplu** [comment]
Awesome!!!! it looks very similar to the changes I'm currently working on, but apparently way much nicer :) I will go deeper into it tomorrow, and let some comments if needed.

File: tf_training_args.py (hunk 1 above)
Comment (jplu) [reply tf_training_args.py:14]: Definitely, each field can have its own enum. I'm also thinking pf reducing the number of properties, or at least find a way to have them in common with the PT Trainer as, indeed, they all can be part of the main TrainingArguments. The big picture would be to have the same conf file doesn't matter if we use the PT or TF trainer.

File: tf_glue_barebones.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+# datasets will do that.
+processor = DataProcessorForSequenceClassification(
+    **vars(data_args), 
+    dataset_name="glue/mrpc",
+    format="tfds",
+    guid="idx",
+    text_a="sentence1",
+    text_b="sentence2",
+    label="label",
+)
```
Comment (jplu) [inline tf_glue_barebones.py:58]: I fully agree. This part (data processor) needs to be way much simplified with Thom's nlp lib.

File: tf_glue_barebones.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+datasets, dataset_info = processor.preprocess_data(args=data_args, tokenizer=tokenizer)
+
+label2id = {label: i for i, label in enumerate(dataset_info.labels)}
+id2label = {i: label for i, label in enumerate(dataset_info.labels)}
```
Comment (jplu) [inline tf_glue_barebones.py:63]: I fully agree. This part (data processor) needs to be way much simplified with Thom's nlp lib.

File: tf_glue_barebones.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    data_dir: str = field(
+        metadata={"help": "The input data dir. Should contain the .tsv files (or other data files) for the task."}
+    )
+    max_seq_length: int = field(
+        default=128,
+        metadata={
+            "help": "The maximum total input sequence length after tokenization. Sequences longer "
+            "than this will be truncated, sequences shorter will be padded."
+        },
+    )
```
Comment (jplu) [inline tf_glue_barebones.py:34]: These fields cannot be part of `TrainingArguments`? Or it doesn't make sense and we should keep them outside?

Comment (jplu) [inline tf_glue_barebones.py:34]: These fields cannot be part of TrainingArguments? Or it doesn't make sense and we should keep them outside?

File: tf_glue_barebones.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+model = TFAutoModelForSequenceClassification.from_pretrained(model_args.model_name_or_path, config=config)
+
+trainer = TFTrainer(
+    model=model,
+    args=training_args,
+    train_dataset=datasets["train"],
+    eval_dataset=datasets["validation"],
+    test_dataset=datasets["test"],
```
Comment (jplu) [inline tf_glue_barebones.py:80]: Should we not reduce this to one single parameter? I mean passing `datasets` itself instead.

File: tf_glue_barebones.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    dataset_info=dataset_info,
+)
+trainer.train()
+trainer.save_model("./save")
```
Comment (jplu) [inline tf_glue_barebones.py:84]: I was thinking if `save_model` cannot be called automatically at the end of the training and then remove this line, as the saving path will be among the trainer arguments. What do you think?

File: trainer_tf_barebones.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+        if self.args.strategy_name == "mirrored":
+            self.strategy = tf.distribute.MirroredStrategy()
+        elif self.args.strategy_name == "onedevice":
+            if len(tf.config.list_physical_devices('GPU')) >= 1:
+                self.strategy = tf.distribute.OneDeviceStrategy(device="/gpu:0")
+            else:
+                self.strategy = tf.distribute.OneDeviceStrategy(device="/cpu:0")
+        else:
+            raise ValueError("The strategy {} does not exists.".format(self.args.strategy_name))
```
Comment (jplu) [inline trainer_tf_barebones.py:64]: This can be simplified by having the same behavior than the `_setup_devices` method in `TrainingArguments`. Is it possible to add it with a `@tensorflow_required` annotation? What do you think?

File: trainer_tf_barebones.py [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+
+        # To conform with Trainer's API we call this from here.
+        # All args should be in the `args` already.
+        self._setup_training()
```
Comment (jplu) [inline trainer_tf_barebones.py:68]: Ok, but I suppose that the `checkpoint_path` and `log_path` parameters will be defined in the `TrainingArguments` and then passed to it?

File: trainer_tf_barebones.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+        if self.args.mode == "classification":
+            for batch in self.test_dataset:
+                y_true.extend(batch[1].numpy().tolist())
+
+            y_pred = np.reshape(np.argmax(results, axis=-1), (-1, 1)).tolist()
+            y_true = list(itertools.chain.from_iterable(y_true))
+            y_pred = list(itertools.chain.from_iterable(y_pred))
+
+            logger.info(classification_report(y_true, y_pred, target_names=self.dataset_info.labels))
```
Comment (jplu) [inline trainer_tf_barebones.py:351]: These two methods are the ones that are task dependent. The first one because of the loss computation and the second one because of the metrics computation. I was thinking, to remove this "dependency" to create a `losses_tf.py` and `metrics_tf.py` files where all the "custom" metrics and losses will be implemented and then becomes fully agnostic because these losses and metrics could be selected in the the training arguments. What do you think?

File: tf_training_args.py (hunk 1 above)
Comment (julien-c) [reply tf_training_args.py:14]: As discussed, I'm fine with keeping a `TFTrainingArguments` for now.

if we need to define a mode here (still not 100% sure) I'd suggest we use "sequence-classification", "token-classification", etc.

File: tf_glue_barebones.py (hunk 5 above)
Comment (julien-c) [reply tf_glue_barebones.py:80]: In PyTorch it felt slightly simpler to do it like that.

File: tf_glue_barebones.py (hunk 6 above)
Comment (julien-c) [reply tf_glue_barebones.py:84]: I would just do `.save_model()`, with an optional override, like in PyTorch:

https://github.com/huggingface/transformers/blob/7c2a32ff888da76b32db839711da2b128698a16c/src/transformers/trainer.py#L411

File: trainer_tf_barebones.py (hunk 7 above)
Comment (julien-c) [reply trainer_tf_barebones.py:64]: Let's add it on the `TFTrainingArguments` then?

File: tf_glue_barebones.py (hunk 4 above)
Comment (julien-c) [reply tf_glue_barebones.py:34]: Better to separate the concerns IMO.

By the way, you should take look at the recently-rewritten PT scripts using Trainer:

- https://github.com/huggingface/transformers/blob/master/examples/run_glue.py
- https://github.com/huggingface/transformers/blob/master/examples/ner/run_ner.py
- https://github.com/huggingface/transformers/blob/master/examples/run_language_modeling.py
- https://github.com/huggingface/transformers/blob/master/examples/run_multiple_choice.py

The run_glue one is slightly different because the `GlueDataTrainingArguments` are defined in another file with a GlueDataset, but we would change that after moving to the new `nlp/datasets`

File: trainer_tf_barebones.py (hunk 8 above)
Comment (julien-c) [reply trainer_tf_barebones.py:68]: For checkpoint_path, in PyTorch we just hardcode that it's going to be `{args.output_dir}/checkpoint-{global_step}`, cf. https://github.com/huggingface/transformers/blob/6af5a54c28c84e248dac2df2ad7b40968fded3da/src/transformers/trainer.py#L361

For log_path, yes, it's called logging_dir: https://github.com/huggingface/transformers/blob/6af5a54c28c84e248dac2df2ad7b40968fded3da/src/transformers/training_args.py#L60

**Comment by jplu** [comment]
Perfect for merging :)
